contract/models/contract.py

```diff
@@ -249,24 +275,16 @@ def recurring_create_invoice(self):
             for line in contract_lines:
                 invoice_values.setdefault('invoice_line_ids', [])
                 invoice_values['invoice_line_ids'].append(
-                    (0, 0, line._prepare_invoice_line(False))
+                    (0, 0, line._prepare_invoice_line(invoice_id=False))
                 )
-            # If no account on the product, the invoice lines account is
-            # taken from the invoice's journal in _onchange_product_id
-            # This code is not in finalize_creation_from_contract because it's
-            # not possible to create an invoice line with no account
-            new_invoice = invoice_model.new(invoice_values)
-            for invoice_line in new_invoice.invoice_line_ids:
-                invoice_line.invoice_id = new_invoice
-                invoice_line._onchange_product_id()
-            invoice_values = new_invoice._convert_to_write(new_invoice._cache)
+            invoice_values = self._finalize_invoice_values(invoice_values)
             invoices_values.append(invoice_values)
             contract_lines._update_recurring_next_date()
         invoices = invoice_model.create(invoices_values)
-        invoices.finalize_creation_from_contract()
+        self._finalize_invoice_creation(invoices)
```

Comment:
I still need two finalization methods: 
 - one before the create (because it's not possible to create invoice lines with no account)
 - one after the create

Reply:
Or you could move finalization and creation in one method that is called after the loop.